### ISSUE TITLE
Too many p-values of 0?

### ISSUE THREAD

**Issue by austinv11** [saezlab/liana#39]
Hello,

Once again I am really grateful for all your work! But I am a little concerned about the p-values I am seeing for permutation tests such as cellphoneDB and cellchat on my data.

I am using a custom resource for my analysis and I am seeing that for cellphonedb and cellchat I am getting >60% of results with a p-value of 0 (even with permutations set to 10,000). This is fairly similar to the results I get if I use call_squidpy (after filtering out the NaNs from squidpy). But, when using the same data in the original CellPhoneDB software externally of liana, the percentage of 0s drops dramatically to ~20%. I have also tried using either the built in consensus resource or the built in cellphone db resource and still get very similar results.

Do you have insight as to why this is happening for me? Or how to investigate further? I am wondering if there are subtle differences in squidpy's (and by extension liana's) implementation of the cellphonedb algorithm. 

I should also note that these p-values 0 seem to be split of values of `0` (integer) and `0.00000` (float) so I wonder if maybe there is some sort of unexpected type coercion/rounding error of the p-values?

**Comment by austinv11** [comment]
Hello, I appreciate your response! Comforting that this is a known phenomenon. 

I suppose it is less of an issue for CellChat since you mention that it is somewhat expected and it provides a secondary `prob` column that I can use for sorting/ranking outside of the p-values.

Though, I should note that we have tried experimenting with the `expr_prop` filter already, and it appears to not really help all too much in our case. Here is some additional information my colleague has generated in case it is useful to you:

When using `expr_prop=0` on our data we get the following results (I am omitting squidpy since liana seems consistent with it):
- CellPhoneDB software calculates 92,826 p-values while liana's cellphonedb calculates 80,109 p-values. But this is not too worrisome since the original cellphonedb calculates p-values without any filtering and there are likely minor discrepancies with the databases used in each software so this is not an issue for us.
- Out of these interactions reported, 76,950 interactions are shared between both implementations (again, the discrepancies in the number of interactions don't really matter) 
- From the shared interactions, 1,934 have 0 as the p-value in both implementations
- The original CellPhoneDB has 1,984 p-values of 0 in total and the liana cellphonedb has 28,346 interactions with a p-value of 0

In short, seems that liana's cellphonedb calculates around 26,000 more p-values of 0 than expected.

Once again, thank you very much for your work and your willingness for discussion! Liana has been really useful to me so far.

**Comment by dbdimitrov** [comment]
Hi @austinv11,

Thanks a lot for looking into this and for raising this point. I will look into it and come back to you.

Daniel

**Comment by dbdimitrov** [comment]
Hi @austinv11,

I believe I might know why this happens.

In short, I believe it's to do with the way that I calculate the p-values. Rather than counting (i.e. knowing the exact number of lr_mean > than the reshuffled means), for the sake of computation speed, I instead use `ecdf` to calculate a percentile for each `lr_mean`, then I use this to estimate the p-value.

I will do some more checks and think if I can circumvent this without doing the actual counting.

**Comment by dbdimitrov** [comment]
I believe there might be a workaround if I z-transform the reshuffled means and then use `pnorm` to calculate the p-values. I will try to implement it in the next update.

In the meantime, similarly to how you use `prob`, one could argue that you can use the `lr_mean` from liana's cpdb to also rank, keeping in mind that it was not intended to be used this way. In fact, I'm working also on a `liana_aggregate` option that would allow to aggregate the scores by the alternatives of each method (e.g. prob, lr_mean, etc). 

As part of LIANA's manuscript, I noticed that the ranks of prob and lr_mean are rather similar. You can simply think of CellChat's prob as a regularized lr_mean weighed by the mediators of interactions, whenever available.

Hope this helps.

**Comment by austinv11** [comment]
Great, thanks for investigating this! I will keep your suggestions in mind. 

Looking forward to the update!

**Comment by austinv11** [comment]
Hey @dbdimitrov 

I appreciate you looking into this. However, I should note that even with the filtering there are still lots of 0s (over 1,000) for my data which makes it difficult to interpret the cpdb results (which is important for me since for my work I need to compare each of the method outputs individually).  Particularly since I am trying to look at the top N predicted interactions with N being less than 1,000 ideally. 

So I assume that I should be looking at lr_mean for ranking in my case? Or should I only use lr_mean to break ties for p-values of 0?

Thank you again for your responses!

**Comment by dbdimitrov** [comment]
Hi @austinv11,

Indeed, I had the same issue when I tried to compare CellChat and CPDB to the remainder of the methods in liana's manuscript. So, as a compromise I used prob/lr_mean after filtering p-values <= 0.05. So, inded in the end I basically used it to break the ties of 0s.

**Comment by austinv11** [comment]
Great, thanks for the info! Feel free to close this issue.

**Comment by schroeme** [comment]
Hi, I have a related question: when I look at the distribution of p-values I get from Liana's CPDB method (pooled for all cluster pairs and all L-R pairs), there are peaks at 0 and 1 with fewer values in between (see below). In agreement with the discussion above, if I return all L-R pair results (`return_all_lrs = True`), the peak at 1 is much larger than the peak at 0, but if I restrict to those with `exp_prop=0.1`, the peak at 0 is larger than the peak at 1. If my understanding of the p-value calculation method is correct, peaks of 1 and 0 with few p-values in between means either the permuted L-R mean is almost either always larger/equal to (for p = 1) or always smaller (p = 0) than the observed L-R mean. Is this understanding correct? Is the reason for this distribution true biological heterogeneity of cell types?

<img width="380" alt="Image" src="https://github.com/user-attachments/assets/464eaf0c-2f19-4f48-8e56-ab186dd670c8" />

Secondly, are cell group labels permuted with replacement (i.e., a cell can be reassigned to its original cluster)? Will this cause any strange behavior if the number of groups/clusters is low, say 4? Is there a minimum number of clusters/groups that CellPhoneDB needs to produce meaningful results?

Thanks!
